src/esm_runscripts/filedicts.py

```
@@ -45,30 +47,112 @@ def __init__(self, full_config, attrs_address):
+
+        }
+        self.names = {
+            "work": pathlib.Path(self["name_in_work"]),
+            "pool": pathlib.Path(self["name_in_pool"]),
         }
         # Allow dot access:
         self.work = self.locations["work"]
         self.pool = self.locations["pool"]
-        self.exp_tree = self.locations["exp_tree"]
-        self.run_tree = self.locations["run_tree"]
+#        self.exp_tree = self.locations["exp_tree"] # TODO: uncomment when lines above are fixed
+#        self.run_tree = self.locations["run_tree"] # TODO: uncomment when lines above are fixed
 
-    def cp(self) -> None:
-        pass
+        # Verbose set to true by default, for now at least
+        self.verbose = full_config.get("general", {}).get("verbose", True)
+
+    def cp(self, source, target) -> None:
+        """
+        Copies the source file or folder to the target path. It changes the name of the
+        target if ``self["name_in_<target>"]`` differs from ``self["name_in_<source>"].
+
+        Parameters
+        ----------
+        source : str
+            String specifying one of the following options: ``"pool"``, ``"work"``,
+            ``"exp_tree"``, ``run_tree``
+        target : str
+            String specifying one of the following options: ``"pool"``, ``"work"``,
+            ``"exp_tree"``, ``run_tree``
+        """
+        # Build target and source paths
+        source_path = self.locations[source].joinpath(self.names[source])
+        target_path = self.locations[target].joinpath(self.names[target])
+
+        # Checks
+        source_path_type = self.check_source_and_target(source_path, target_path)
+
+        # Actual copy
+        if source_path_type == "dir":
+            copy_func = shutil.copytree
+        else:
+            copy_func = shutil.copy2
+        try:
+            copy_func(source_path, target_path)
+        except Exception:
+            user_error("Filedict Error", f"Unable to copy {source_path} to {target_path}")
```

I would not use `user_error` here. It is a very strong function because it aborts the whole executable. I think it is better to raise an exception and let the driver routine to give the final decision.

---

Agreed